### PR TITLE
Lint with ruff

### PR DESCRIPTION
**Pull request by jhamman**
This switches our primary linter to Ruff. As adervertised, Ruff is very fast. Plust we get the benefit of using a single tool that combines the previous functionality of pyflakes, isort, and pyupgrade.

- [x] Closes https://twitter.com/TEGNicholasCode/status/1613226956887056385
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

cc @max-sixty, @TomNicholas

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -69,3 +69,35 @@ module = [
...
+# E203: whitespace before ':' - doesn't work well with black
+# E402: module level import not at top of file
+# E501: line too long - let black worry about that
+# E731: do not assign a lambda expression, use a def
+# W503: line break before binary operator
+ignore = [
+    "E402",
+    "E501",
+    "E731",
+]
```
Comment (jhamman) [inline pyproject.toml:89]: E203 and W503 didn't make the ignore list using `flake8-to-ruff`

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -14,22 +14,13 @@ repos:
...
-  - repo: https://github.com/PyCQA/isort
-    rev: 5.11.4
-    hooks:
-      - id: isort
-  - repo: https://github.com/asottile/pyupgrade
-    rev: v3.3.1
-    hooks:
-      - id: pyupgrade
-        args:
-          - "--py38-plus"
```
Comment (jhamman) [inline .pre-commit-config.yaml:32]: It doesn't look like this is supported by Ruff. We'll need to walk back the type changes for py38...

File: xarray/tests/test_accessor_str.py [hunk 3]
```diff
@@ -104,7 +104,7 @@ def test_contains(dtype) -> None:
     values = xr.DataArray(["Foo", "xYz", "fOOomMm__fOo", "MMM_"]).astype(dtype)
 
     # case insensitive using regex
-    pat = values.dtype.type("FOO|mmm")
+    pat = str
```
Comment (jhamman) [inline xarray/tests/test_accessor_str.py:107]: I'm not sure what this was doing before I'm fairly sure we don't want this change

File: .pre-commit-config.yaml (hunk 2 above)
Comment (charliermarsh) [reply .pre-commit-config.yaml:32]: I think we do support this! Unless I'm misunderstanding. But we use something closer to Black's API, like:

```py
[tool.ruff]
# Always generate Python 3.7-compatible code.
target-version = "py38"
```

File: xarray/tests/test_accessor_str.py (hunk 3 above)
Comment (Illviljan) [reply xarray/tests/test_accessor_str.py:107]:
```python
values = xr.DataArray(["Foo", "xYz", "fOOomMm__fOo", "MMM_"]).astype(np.str_)
values.dtype.type("FOO|mmm")
Out[44]: 'FOO|mmm'

values = xr.DataArray(["Foo", "xYz", "fOOomMm__fOo", "MMM_"]).astype(np.bytes_)
values.dtype.type("FOO|mmm")
Out[46]: b'FOO|mmm'
```

Looks like a bug. Probably trying to convert `type("FOO|mmm")` to `str` but doesn't take into account that it's a method and not the python function.

Comment (charliermarsh) [reply xarray/tests/test_accessor_str.py:107]: Hmm, I think this was fixed in https://github.com/charliermarsh/ruff/pull/1962 (which went out in `v0.0.226`).

Comment (Illviljan) [reply xarray/tests/test_accessor_str.py:107]: I think this ran with 0.0.226, right @jhamman ?

```
https://github.com/charliermarsh/ruff-pre-commit@v0.0.226 (54.6KiB) (cached)
  sha: 6aa52c1faae70c9c500f47c60b66e00214d48f52
  envs:
  - python@python3 (10.4MiB) (cached)
```

Comment (jhamman) [reply xarray/tests/test_accessor_str.py:107]: Actually, I think I ran `ruff --fix` with 0.0.146. I'll revert, update and try again.

**Comment by jhamman** [comment]
I plan to leave this as a WIP PR until #7461 is merged. There is a bit of overlap and dropping Python 3.8 will make this much easier to finish up.

File: xarray/tests/test_distributed.py [hunk 4]
```diff
@@ -19,35 +18,34 @@
...
+)
+from xarray.tests.test_backends import (
+    ON_WINDOWS,
+    create_tmp_file,
+    create_tmp_geotiff,
+    open_example_dataset,
 )
+from xarray.tests.test_dataset import create_test_data
 
 # this is to stop isort throwing errors. May have been easier to just use
```
Comment (dcherian) [inline xarray/tests/test_distributed.py:50]: Could delete these comments and move the `da` import up to Line 19 without the `importorskip`

very minor comment though, we can just leave it as-is

**Review by dcherian — APPROVED** [review]

File: xarray/tests/test_distributed.py (hunk 4 above)
Comment (jhamman) [reply xarray/tests/test_distributed.py]: you got it!

**Pull request merged**